### ISSUE TITLE
io:fread on OTP 26 is much slower than on OTP 25

### ISSUE THREAD

**Issue by susurri** [erlang/otp#7924]
**Describe the bug**
io:fread on OTP 26 is much slower than on OTP 25

**To Reproduce**
Main.erl
```
-module('Main').
-export([main/1]).

main(_) ->
    N = 500000,
    _List = io:fread("", (lists:flatten(lists:duplicate(N, "~d")))),
    halt().
```

input.txt is a 500000 lines of text file.
```
500000
500000
500000
...
```
`erlc Main.erl`
`/usr/bin/time erl -noshell -run Main main run < input.txt`

OTP 26.1.2
```
15.56user 1.65system 0:06.24elapsed 275%CPU (0avgtext+0avgdata 257764maxresident)k
0inputs+0outputs (0major+49661minor)pagefaults 0swaps
```

OTP 25.3.2.7
```
1.21user 0.23system 0:00.88elapsed 164%CPU (0avgtext+0avgdata 259308maxresident)k
0inputs+0outputs (0major+45920minor)pagefaults 0swaps
```

**Expected behavior**
It runs on OTP 26 as fast as on OTP 25.

**Affected versions**
26.1.2 (probably 26.*)

**Additional context**

**Comment by garazdawi** [comment]
Most of the regression seems to have been introduced in 26.0.2 when I was fixing bugs in latin1/unicode handling.

Please try #7933 and see if that solves the original issue for you.

**Comment by susurri** [comment]
```
4.41user 0.66system 0:01.76elapsed 288%CPU (0avgtext+0avgdata 290312maxresident)k
0inputs+0outputs (0major+60043minor)pagefaults 0swaps
```
It is much better but still slower than OTP 25.

**Comment by garazdawi** [comment]
We do more validation of the input data in order to give better error messages, so some slowdown is to be expected. I will see if I can make it a bit faster by moving more validation to the edges.

**Comment by garazdawi** [comment]
I did not find a good way to do it yet, but I merged the fix in #7933 to be part of Erlang/OTP 26.3. I'll try to keep this in mind next time I'm in group changing things.
